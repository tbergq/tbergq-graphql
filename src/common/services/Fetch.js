// @flow

import fetch from 'node-fetch';

export default async (url: string, options: Object = {}) => {
  try {
    const response = await fetch(url, options);

    const json = await response.json();
    return json;
  } catch (err) {
    console.log(err);
    throw err;
  }
};
