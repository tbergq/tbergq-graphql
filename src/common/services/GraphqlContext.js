// @flow

import DataLoader from 'dataloader';
import jwt from 'jsonwebtoken';
import { config } from 'dotenv';

import SearchTvShowLoader, {
  type TvShow,
} from '../../tvhelper/dataloaders/SearchTvShowLoader';
import TvDetailLoader from '../../tvhelper/dataloaders/TvDetailLoader';
import TvHelperUserLoader, {
  type UserType as TvHelperUser,
} from '../../tvhelper/dataloaders/UserLoader';
import FavoritesLoader, {
  type Favorites,
} from '../../tvhelper/dataloaders/FavoritesLoader';
import TvShowEpisodesLoader, {
  type Episode,
} from '../../tvhelper/dataloaders/TvShowEpisodesLoader';
import FavoriteLoader, {
  type FavoriteArgs,
} from '../../tvhelper/dataloaders/FavoriteLoader';
import EpiosodeWatchedLoader, {
  type EpisodeWatched,
} from '../../tvhelper/dataloaders/EpisodeWatchedLoader';
import EpisodeLoader from '../../tvhelper/dataloaders/EpisodeLoader';
import ProgramsLoader, {
  type ProgramsParams,
  type Programs,
} from '../../trainingjournal/programs/dataloaders/ProgramsLoader';

config();

const { JWT_SECRET } = process.env;

export type LoggedInUser = {|
  +id: string,
  +username: string,
|};

export type GraphqlContextType = {|
  +user: ?LoggedInUser,
  +rawToken: ?string,
  +dataLoader: {|
    +tvhelper: {|
      +searchTvShow: DataLoader<string, TvShow[]>,
      +tvDetail: DataLoader<string, TvShow>,
      +user: DataLoader<string, TvHelperUser>,
      +favorites: DataLoader<string, Favorites[]>,
      +episodes: DataLoader<string, Episode[]>,
      +favorite: DataLoader<FavoriteArgs, Favorites>,
      +episodeWatched: DataLoader<number, EpisodeWatched[]>,
      +episode: DataLoader<string, Episode>,
    |},
    +trainingjournal: {|
      +programs: DataLoader<ProgramsParams, Programs>,
    |},
  |},
|};

const decodeToken = (token: ?string) => {
  try {
    return token == null ? null : jwt.verify(token, JWT_SECRET);
  } catch (err) {
    return null;
  }
};

export default function createContext(token: ?string) {
  const user = decodeToken(token);
  return {
    user,
    rawToken: token,
    dataLoader: {
      tvhelper: {
        searchTvShow: SearchTvShowLoader(),
        tvDetail: TvDetailLoader(),
        user: TvHelperUserLoader(),
        favorites: FavoritesLoader(),
        episodes: TvShowEpisodesLoader(),
        favorite: FavoriteLoader(),
        episodeWatched: EpiosodeWatchedLoader(user),
        episode: EpisodeLoader(),
      },
      trainingjournal: {
        programs: ProgramsLoader(),
      },
    },
  };
}
