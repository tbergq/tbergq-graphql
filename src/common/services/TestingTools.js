// @flow

import { graphql as originalGraphQL } from 'graphql';
import schema from '../../Schema';
import createContext from './GraphqlContext';

export const graphql = async (
  query: string,
  variables: ?Object,
): Promise<Object> =>
  originalGraphQL(schema, query, null, createContext(), variables);
