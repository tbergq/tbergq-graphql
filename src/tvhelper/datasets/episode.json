{
    "id": 1471697,
    "url": "http://www.tvmaze.com/episodes/1471697/the-100-5x13-damocles-part-two",
    "name": "Damocles â€“ Part Two",
    "season": 5,
    "number": 13,
    "airdate": "2018-08-07",
    "airtime": "20:00",
    "airstamp": "2018-08-08T00:00:00+00:00",
    "runtime": 60,
    "image": {
        "medium": "http://static.tvmaze.com/uploads/images/medium_landscape/163/409348.jpg",
        "original": "http://static.tvmaze.com/uploads/images/original_untouched/163/409348.jpg"
    },
    "summary": "<p>Clarke and her friends must risk everything to fight one last battle for survival, only to glimpse an even darker threat to the last living valley on Earth.</p>",
    "_links": {
        "self": {
            "href": "http://api.tvmaze.com/episodes/1471697"
        }
    }
}
