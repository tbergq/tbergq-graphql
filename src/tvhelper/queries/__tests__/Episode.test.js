// @flow

import { toGlobalId } from 'graphql-relay';

import { graphql } from '../../../common/services/TestingTools';
import episode from '../../datasets/episode.json';

const query = `
  query episode($id: ID!) {
    episode(id: $id) {
      id
      name
      image {
        medium
        original
      }
      season
      number
      seasonAndNumber
      airdate
      summary(stripTags: false)
    }
  }
  `;

it('works', async () => {
  // $FlowFixMe
  fetch.mockResponse(JSON.stringify(episode));
  expect(await graphql(query, { id: toGlobalId('episode', '1471697') })).toMatchSnapshot();
});
