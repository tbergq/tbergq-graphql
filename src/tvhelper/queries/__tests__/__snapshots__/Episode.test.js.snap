// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`works 1`] = `
Object {
  "data": Object {
    "episode": Object {
      "airdate": "2018-08-07",
      "id": "ZXBpc29kZToxNDcxNjk3",
      "image": Object {
        "medium": "http://static.tvmaze.com/uploads/images/medium_landscape/163/409348.jpg",
        "original": "http://static.tvmaze.com/uploads/images/original_untouched/163/409348.jpg",
      },
      "name": "Damocles â€“ Part Two",
      "number": 13,
      "season": 5,
      "seasonAndNumber": "S05E13",
      "summary": "<p>Clarke and her friends must risk everything to fight one last battle for survival, only to glimpse an even darker threat to the last living valley on Earth.</p>",
    },
  },
}
`;
